### PR TITLE
[Inference API] Prevent inference endpoints from being deleted if they are referenced by semantic text

### PR DESCRIPTION
**Pull request by maxhniebergall**
Following on https://github.com/elastic/elasticsearch/pull/109123

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    private static final Logger logger = LogManager.getLogger(SemanticTextInfoExtractor.class);
+
+    public static Set<String> extractSemanticTextFields(Metadata metadata, Set<String> endpointIds) {
+        assert endpointIds.isEmpty() == false;
+        assert metadata != null;
+
+        Set<String> referenceIndices = new HashSet<>();
+
+        Map<String, IndexMetadata> indices = metadata.indices();
+        logger.error("starting extract");
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java]: Debug logging statement we need to remove?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        indices.forEach((indexName, indexMetadata) -> {
+            if (indexMetadata.getInferenceFields() != null) {
+                Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
+                logger.error("indexName: " + indexName);
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java]: Another debug logging statement

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                if (inferenceFields.entrySet()
+                    .stream()
+                    .anyMatch(
+                        entry -> entry.getValue().getInferenceId() != null && endpointIds.contains(entry.getValue().getInferenceId())
+                    )) {
+                    logger.error("indexName: " + indexName + "inner");
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java]: Another debug logging statement

File: x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceBaseRestTest.java [hunk 4]
```diff
@@ -126,6 +126,25 @@ protected void deleteModel(String modelId, TaskType taskType) throws IOException
...
 
+    protected void putSemanticText(String endpointId, String indexName) throws IOException {
+        var request = new Request("PUT", Strings.format("%s", indexName));
+        String body = Strings.format("""
+            {
+                "mappings": {
+                "properties": {
+                    "inference_field": {
+                        "type": "semantic_text",
+                            "inference_id": "%s"
```
Comment (Mikep86) [inline x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceBaseRestTest.java:137]: Nit: indentation is off here

File: x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java [hunk 5]
```diff
@@ -146,4 +146,35 @@ public void testDeleteEndpointWhileReferencedByPipeline() throws IOException {
...
+        var pipelineId = "pipeline_referencing_model";
+        putPipeline(pipelineId, endpointId);
+
+        {
+            var e = expectThrows(ResponseException.class, () -> deleteModel(endpointId));
+            assertThat(
+                e.getMessage(),
+                containsString(
+                    " is referenced by `SemanticText` field(s) and cannot be deleted. "
+                        + "Use `force` to delete it anyway, or use `dry_run` to list the `SemanticText` field(s) that reference it."
```
Comment (Mikep86) [inline x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java:161]: We should use `semantic_text` here to match the name of the field type

File: x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java [hunk 6]
```diff
@@ -146,4 +146,35 @@ public void testDeleteEndpointWhileReferencedByPipeline() throws IOException {
         }
         deletePipeline(pipelineId);
     }
+
+    public void testDeleteEndpointWhileReferencedBySemanticText() throws IOException {
+        String endpointId = "endpoint_referenced_by_pipeline";
+        putModel(endpointId, mockSparseServiceModelConfig(), TaskType.SPARSE_EMBEDDING);
+        putSemanticText(endpointId, randomAlphaOfLength(10));
+        var pipelineId = "pipeline_referencing_model";
+        putPipeline(pipelineId, endpointId);
```
Comment (Mikep86) [inline x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java]: Could we remove the pipeline from this test to keep it more focused?

File: x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java [hunk 7]
```diff
@@ -146,4 +146,35 @@ public void testDeleteEndpointWhileReferencedByPipeline() throws IOException {
...
+                    " is referenced by `SemanticText` field(s) and cannot be deleted. "
+                        + "Use `force` to delete it anyway, or use `dry_run` to list the `SemanticText` field(s) that reference it."
+                )
+            );
+        }
+        {
+            var response = deleteModel(endpointId, "dry_run=true");
+            var entityString = EntityUtils.toString(response.getEntity());
+            assertThat(entityString, containsString(pipelineId));
+            assertThat(entityString, containsString("\"acknowledged\":false"));
```
Comment (Mikep86) [inline x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java:168]: Can we test that the index that contains the semantic text field is reported here?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 8]
```diff
@@ -87,20 +89,31 @@ protected void masterOperation(
                 listener.onFailure(InferenceExceptions.mismatchedTaskTypeException(request.getTaskType(), unparsedModel.taskType()));
                 return;
             }
-
+            logger.error("Stopping inference endpoint {}", request.getInferenceEndpointId());
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 9]
```diff
@@ -87,20 +89,31 @@ protected void masterOperation(
...
             if (request.isDryRun()) {
-                masterListener.onResponse(
-                    new DeleteInferenceEndpointAction.Response(
-                        false,
-                        InferenceProcessorInfoExtractor.pipelineIdsForResource(state, Set.of(request.getInferenceEndpointId()))
-                    )
+                logger.error("Stopping inference endpoint {} with dry run 1", request.getInferenceEndpointId());
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 10]
```diff
@@ -87,20 +89,31 @@ protected void masterOperation(
...
+
+                Set<String> pipelines = InferenceProcessorInfoExtractor.pipelineIdsForResource(
+                    state,
+                    Set.of(request.getInferenceEndpointId())
                 );
+                logger.error("Stopping inference endpoint {} with dry run 2", request.getInferenceEndpointId());
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 11]
```diff
@@ -87,20 +89,31 @@ protected void masterOperation(
...
+
+                Set<String> indexesReferencedBySemanticText = extractSemanticTextFields(
+                    state.getMetadata(),
+                    Set.of(request.getInferenceEndpointId())
+                );
+                logger.error("Stopping inference endpoint {} with dry run 3", request.getInferenceEndpointId());
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 12]
```diff
@@ -87,20 +89,31 @@ protected void masterOperation(
...
+
+                masterListener.onResponse(new DeleteInferenceEndpointAction.Response(false, pipelines, indexesReferencedBySemanticText));
                 return;
             } else if (request.isForceDelete() == false
-                && endpointIsReferencedInPipelines(state, request.getInferenceEndpointId(), listener)) {
+                && endpointIsReferenceInPipelinesOrSemanticText(state, request.getInferenceEndpointId(), listener)) {
                     return;
                 }
 
+            logger.error("Stopping inference endpoint {} after check", request.getInferenceEndpointId());
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 13]
```diff
@@ -126,9 +139,41 @@ && endpointIsReferencedInPipelines(state, request.getInferenceEndpointId(), list
...
+                )
+            );
+    }
+
+    private static boolean endpointIsReferenceInPipelinesOrSemanticText(
+        final ClusterState state,
+        final String inferenceEndpointId,
+        ActionListener<Boolean> listener
+    ) {
+        logger.error("Stopping inference endpoint -- check");
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java]: Debug logging statement

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java [hunk 14]
```diff
@@ -126,9 +139,41 @@ && endpointIsReferencedInPipelines(state, request.getInferenceEndpointId(), list
...
+        final String inferenceEndpointId,
+        ActionListener<Boolean> listener
+    ) {
+        if (extractSemanticTextFields(state.getMetadata(), Set.of(inferenceEndpointId)).isEmpty() == false) {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    "Inference endpoint "
+                        + inferenceEndpointId
+                        + " is referenced by `SemanticText` field(s) and cannot be deleted. "
+                        + "Use `force` to delete it anyway, or use `dry_run` to list the `SemanticText` field(s) that reference it.",
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java:163]: Let's work on the language here. We may want to be more generic and just say that the inference endpoint is referenced in index mappings to be forward-looking. I'm thinking of a future where the `sparse_vector` or `dense_vector` field types have an `inference_id` param, which may happen.

**Review by Mikep86 — COMMENTED** [review]
Thanks for the quick work here! I think we should take a second pass and consider whether we want to call out semantic text specifically in method names and error messages. We may add `inference_id` to other field types in the future (such as `dense_vector` or `sparse_vector`).

File: x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java [hunk 15]
```diff
@@ -146,4 +146,32 @@ public void testDeleteEndpointWhileReferencedByPipeline() throws IOException {
         }
         deletePipeline(pipelineId);
     }
+
+    public void testDeleteEndpointWhileReferencedBySemanticText() throws IOException {
+        String endpointId = "endpoint_referenced_by_pipeline";
```
Comment (carlosdelest) [inline x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java:151]:
```suggestion
        String endpointId = "endpoint_referenced_by_semantic_text";
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java (hunk 14 above)
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportDeleteInferenceEndpointAction.java:163]: ++ . Also, I think we're returning the indices that contain it, and not the actual fields.

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/50_semantic_text_query_inference_endpoint_changes.yml [hunk 16]
```diff
@@ -78,6 +78,9 @@ setup:
         refresh: true
 ---
 "sparse_embedding changed to text_embedding":
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/50_semantic_text_query_inference_endpoint_changes.yml:80]: @carlosdelest, I am having trouble understanding what these tests are doing and why they are failing. The error message they were giving was:
```     
   Caused by:
        java.lang.AssertionError: expected [2xx] status code but api [inference.delete] returned [409 Conflict] [{"error":{"root_cause":[{"type":"status_exception","reason":" Inference endpoint sparse-inference-id is being used in the mapping for indexes: [test-sparse-index]. Ensure that no index mappings are using this inference endpoint, or use force to ignore this warning and delete the inference endpoint.",
```
I am not sure why these tests are attempting to delete an inference endpoint at the begging, should that be happening at the end the test where the endpoint was created?

Comment (Mikep86) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/50_semantic_text_query_inference_endpoint_changes.yml:80]: I wrote this test, it's testing when the user replaces an inference endpoint that uses a sparse embedding model with one that uses a text embedding model. We should be able to fix it by using `force` with the `inference.delete` call.

Comment (maxhniebergall) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/50_semantic_text_query_inference_endpoint_changes.yml:80]: thanks Mike!

**Pull request merged**